### PR TITLE
added simple json websocket protocol

### PR DESCRIPTION
**Pull request by virtustate**

### CODE REVIEW

File: src/local_options [hunk 1]
```diff
@@ -25,7 +25,7 @@
 #undef NO_ENVIRONMENT
 #undef NO_WIZARDS
 #undef NO_LIGHT
-#define OLD_ED
+#undef OLD_ED
```
Comment (thefallentree) [inline src/local_options:28]: pls don't submit this change

File: src/net/websocket.cc [hunk 2]
```diff
@@ -11,11 +11,13 @@
 enum PROTOCOL_ID {
   WS_HTTP = 0,
   WS_ASCII = PROTOCOL_WS_ASCII,
+  WS_JSON = PROTOCOL_WS_JSON,
 };
 
 static struct lws_protocols protocols[] = {
     {"http", lws_callback_http_dummy, 0, 0, WS_HTTP},
     {"ascii", ws_ascii_callback, sizeof(struct ws_ascii_session), 4096, WS_ASCII, NULL, 0},
+    {"json", ws_ascii_callback, sizeof(struct ws_ascii_session), 4096, WS_JSON, NULL, 0},
```
Comment (thefallentree) [inline src/net/websocket.cc:20]: the callbacks are wrong , also the sesion struct name

File: src/net/ws_ascii.cc [hunk 3]
```diff
@@ -162,7 +162,7 @@ int ws_ascii_callback(struct lws *wsi, enum lws_callback_reasons reason, void *u
 }
 
 void ws_ascii_send(struct lws *wsi, const char *data, size_t len) {
-  DEBUG_CHECK(lws_get_protocol(wsi)->id != PROTOCOL_WS_ASCII, "wrong protocol!");
+  // DEBUG_CHECK(lws_get_protocol(wsi)->id != PROTOCOL_WS_ASCII, "wrong protocol!");
```
Comment (thefallentree) [inline src/net/ws_ascii.cc:165]: should revert this

File: src/net/websocket.cc (hunk 2 above)
Comment (virtustate) [reply src/net/websocket.cc:20]: Thanks for the review. I backed out the local_options change, put back the DEBUG_CHECK in ws_ascii.cc including both ascii/json. I'm just trying to add an alternate name to the existing ascii websocket protocol (json) that simply results in adding {"message": ...} around what it already sends to the client but use the exact same ws_asscii_callback/session code. Should I really duplicate this callback/structure code with another name? Thanks for the help!

**Comment by thefallentree** [comment]
Yeah, the incidental reuse will make things much worse down the road, I
think we should separate them out.

How ever, instead of just do an basic json split by line, could you
consider implementing something like socket.io ? This also would require
some efun and apply  to send and receive Typed messages, this would enable
client to process json based message directly and would be very useful I
think.

**Comment by thefallentree** [comment]
https://medium.com/@tomokazukozuma/how-to-use-socket-io-protocol-42fa6dd93b0e

here is something that you can reference. It is not the best clearly
documented protocol, but it beats inventing our own if possible, that way
we can enable client to use existing libraries to build on top of it.

**Comment by virtustate** [comment]
Thank you, interesting suggestions in the same general direction I'm thinking. https://github.com/virtustate/FluffSoulClient is a dead simple working angular prototype client with the output in the developer console using https://rxjs-dev.firebaseapp.com/api/webSocket/webSocket. The thought is that everything beyond the the outer driver "message" json wrapper happens in the mudlib and angular client. For example, the room mudlib object outputs something like 
{"room": {"short": "short description, "long": long description", 
"exits":  {"name": "exit1", "path": "path1"}, {"name": "exit1", "path": "path1"},
"items": {...}
...
}
which the angular client unpacks and visually "renders" down its object hierarchy. 

This should mean no additional changes to the driver to support this. It should also allow incremental paired mudlib/angular client implementation by base object with anything not yet json implemented dumped as the usual text output. The outer driver json "message" wrapper may not even be needed but it's a little easier for the client to assume any incoming message is json at least at the top level.

Of course, time is the bottleneck. I start a new full time job tomorrow and will have very little time.

**Pull request closed without merging**